ksqldb-engine/src/main/java/io/confluent/ksql/ddl/commands/DdlCommandExec.java

```diff
@@ -127,6 +131,81 @@ public DdlCommandResult executeDropType(final DropTypeCommand dropType) {
           ? new DdlCommandResult(true, "Dropped type '" + typeName + "'")
           : new DdlCommandResult(true, "Type '" + typeName + "' does not exist");
     }
+
+    @Override
+    public DdlCommandResult executeAlterSource(final AlterSourceCommand alterSource) {
+      final DataSource dataSource = metaStore.getSource(alterSource.getSourceName());
+
+      if (dataSource == null) {
+        throw new KsqlException(
+            "Source " + alterSource.getSourceName().text()
+                + " does not exist."
+        );
+      }
+
+      if (!dataSource.getDataSourceType().getKsqlType().equals(alterSource.getKsqlType())) {
+        throw new KsqlException(String.format(
+            "Incompatible data source type is %s, but statement was ALTER %s",
+            dataSource.getDataSourceType().getKsqlType(),
+            alterSource.getKsqlType()
+        ));
+      }
+
+      if (dataSource.isCasTarget()) {
+        throw new KsqlException(String.format(
+            "ALTER command is not supported for CREATE ... AS statements."
+        ));
+      }
+
+      final LogicalSchema newSchema = dataSource.getSchema()
+          .asBuilder()
+          .valueColumns(alterSource.getNewColumns().columns())
+          .build();
+
+      if (alterSource.getKsqlType().equals(DataSourceType.KSTREAM.getKsqlType())) {
+        return alterStream(dataSource, newSchema);
+      } else {
+        return alterTable(dataSource, newSchema);
+      }
+    }
+
+    private DdlCommandResult alterStream(
+        final DataSource dataSource,
+        final LogicalSchema newSchema
+    ) {
+      final KsqlStream<?> ksqlStream = new KsqlStream<>(
```

Comment:
instead of constructing this here, we should add a method to `KsqlStream` labeled `withSchema`. This ensures that if anyone ever modifies `KsqlStream` and adds new arguments it won't accidentally be missed when we copy it here (same for `KsqlTable`).

also, it could make it cleaner if you just added it to `DataSource` API, then you don't even need to check which type it is :)